src/cache/disk.c

```diff
@@ -0,0 +1,459 @@
+#define _CRT_SECURE_NO_WARNINGS
+#include <errno.h>
+#include <fcntl.h>
+#include <stdlib.h>
+
+#include "private_config.h"
+
+#ifdef _WIN32
+#define PATH_MAX 255
+
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+
+#include <process.h>
+#include <direct.h>
+#include <io.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+
+struct timezone;
+
+struct timeval {
+  long tv_sec;
+  long tv_usec;
+} timeval;
+
+static int gettimeofday(struct timeval *tp, struct timezone *tzp) {
+  /*
+   * Note: some broken versions only have 8 trailing zero's, the
+   * correct epoch has 9 trailing zero's This magic number is the
+   * number of 100 nanosecond intervals since January 1, 1601 (UTC)
+   * until 00:00:00 January 1, 1970
+   */
+  static const uint64_t EPOCH = ((uint64_t)116444736000000000ULL);
+
+  SYSTEMTIME system_time;
+  FILETIME file_time;
+  uint64_t time;
+
+  GetSystemTime(&system_time);
+  SystemTimeToFileTime(&system_time, &file_time);
+  time = ((uint64_t)file_time.dwLowDateTime);
+  time += ((uint64_t)file_time.dwHighDateTime) << 32;
+
+  tp->tv_sec = (long)((time - EPOCH) / 10000000L);
+  tp->tv_usec = (long)(system_time.wMilliseconds * 1000);
+  return 0;
+}
+
+#define open _open
+#define unlink _unlink
+#define mkdir(p, f) _mkdir(p)
+#define close _close
+#define strdup _strdup
+#define lstat _stat64
+#define fstat _fstat64
+#define stat __stat64
+
+#else
+#define PATH_MAX 1024
+#include <unistd.h>
+#include <sys/time.h>
+#include <sys/stat.h>
+
+#define O_BINARY 0
+#define _setmode(a, b)
+
+#endif
+
+
+#include "cache.h"
+#include "util/skein.h"
+
+#define HEXP_LEN (128 + 2)
+
+typedef struct _disk_cache {
+  cache c;
+  cache * mem;
+  kwrite_fn kwrite;
+  vwrite_fn vwrite;
+  kread_fn kread;
+  vread_fn vread;
+  const char *dirp;
+} disk_cache;
+
+
+static unsigned long long ntohull(const char *_in) {
```

Comment:
Like ntohl but for unsigned long long (64 bits).

It converts from network byte order to host byte order.